File: app/src/main/java/blagodarie/rating/ui/user/abilities/AbilitiesFragment.java
package blagodarie.rating.ui.user.abilities;

import android.accounts.Account;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import org.jetbrains.annotations.NotNull;

import java.util.UUID;

import blagodarie.rating.AppExecutors;
import blagodarie.rating.databinding.AbilitiesFragmentBinding;
import blagodarie.rating.model.entities.Ability;
import blagodarie.rating.repository.AsyncRepository;
import blagodarie.rating.repository.AsyncServerRepository;
import blagodarie.rating.ui.user.keys.KeysFragment;
import blagodarie.rating.ui.user.profile.ProfileFragmentArgs;

public final class AbilitiesFragment
        extends Fragment {

    public interface FragmentCommunicator {
        void toEditAbility (@Nullable final Ability ability);
    }

    public interface UserActionListener {
        void onAddAbilityClick ();
    }

    private static final String TAG = AbilitiesFragment.class.getSimpleName();

    private AbilitiesViewModel mViewModel;

    private AbilitiesFragmentBinding mBinding;

    private AbilitiesAdapter mAbilitiesAdapter;

    private Account mAccount;

    private UUID mUserId;

    private final AsyncRepository mAsyncRepository = new AsyncServerRepository(AppExecutors.getInstance().networkIO(), AppExecutors.getInstance().mainThread());

    private final UserActionListener mUserActionListener = new UserActionListener() {
        @Override
        public void onAddAbilityClick () {
            mFragmentCommunicator.toEditAbility(null);
        }
    };

    private FragmentCommunicator mFragmentCommunicator;

    @NotNull
    @Override
    public View onCreateView (
            @NonNull final LayoutInflater inflater,
            @Nullable final ViewGroup container,
            @Nullable final Bundle savedInstanceState
    ) {
        Log.d(TAG, "onCreateView");
        initBinding(inflater, container);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated (
            @NonNull final View view,
            @Nullable final Bundle savedInstanceState
    ) {
        Log.d(TAG, "onViewCreated");
        super.onViewCreated(view, savedInstanceState);

        final ProfileFragmentArgs args = ProfileFragmentArgs.fromBundle(requireArguments());

        mUserId = args.getUserId();
        mAccount = args.getAccount();
    }

    @Override
    public void onActivityCreated (@Nullable final Bundle savedInstanceState) {
        Log.d(TAG, "onActivityCreated");
        super.onActivityCreated(savedInstanceState);
        try {
            mFragmentCommunicator = (FragmentCommunicator) requireActivity();
        } catch (ClassCastException e) {
            Log.e(TAG, requireActivity().getClass().getName() + " must implement " + KeysFragment.FragmentCommunicator.class.getName());
            throw new ClassCastException(requireActivity().getClass().getName() + " must implement " + KeysFragment.FragmentCommunicator.class.getName());
        }

        initOperationsAdapter();
        initViewModel();
        setupBinding();
    }

    @Override
    public void onStart () {
        Log.d(TAG, "onStart");
        super.onStart();
        refreshAbilities();
    }

    @Override
    public void onDestroy () {
        Log.d(TAG, "onDestroy");
        super.onDestroy();
        mBinding = null;
    }

    private void initBinding (
            @NonNull final LayoutInflater inflater,
            @Nullable final ViewGroup container
    ) {
        Log.d(TAG, "initBinding");
        mBinding = AbilitiesFragmentBinding.inflate(inflater, container, false);
    }

    private void initViewModel () {
        Log.d(TAG, "initViewModel");
        mViewModel = new ViewModelProvider(requireActivity()).get(AbilitiesViewModel.class);
        mViewModel.isOwnProfile().set(mAccount != null && mAccount.name.equals(mUserId.toString()));
    }

    private void refreshAbilities () {
        Log.d(TAG, "refreshAbilities");
        mViewModel.setAbilities(mAsyncRepository.getLiveDataPagedListFromDataSource(new AbilitiesDataSource.AbilitiesDataSourceFactory(mUserId)));
        mViewModel.getAbilities().observe(requireActivity(), mAbilitiesAdapter::submitList);
    }

    private void setupBinding () {
        mBinding.setViewModel(mViewModel);
        mBinding.setUserActionListener(mUserActionListener);
        mBinding.rvAbilities.setLayoutManager(new LinearLayoutManager(requireContext()));
        mBinding.rvAbilities.setAdapter(mAbilitiesAdapter);
        mBinding.srlRefreshProfileInfo.setOnRefreshListener(() -> {
            mViewModel.getDownloadInProgress().set(true);
            refreshAbilities();
            mViewModel.getDownloadInProgress().set(false);
        });
    }

    private void initOperationsAdapter () {
        mAbilitiesAdapter = new AbilitiesAdapter();
    }

}

File: app/src/main/java/blagodarie/rating/ui/AbilitiesFragment.kt
package blagodarie.rating.ui

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavDirections
import androidx.navigation.fragment.NavHostFragment
import androidx.paging.PagedList
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import blagodarie.rating.AppExecutors
import blagodarie.rating.databinding.AbilitiesFragmentBinding
import blagodarie.rating.model.IAbility
import blagodarie.rating.repository.AsyncServerRepository
import blagodarie.rating.ui.user.abilities.AbilitiesAdapter
import blagodarie.rating.ui.user.abilities.AbilitiesDataSource
import java.util.*

class AbilitiesFragment : Fragment() {

    interface UserActionListener {
        fun onAddAbilityClick()
    }

    companion object {
        private val TAG = AbilitiesFragment::class.java.simpleName
    }

    private lateinit var mViewModel: AbilitiesViewModel

    private lateinit var mBinding: AbilitiesFragmentBinding

    private lateinit var mAbilitiesAdapter: AbilitiesAdapter

    private lateinit var mUserId: UUID

    private val mAsyncRepository = AsyncServerRepository(AppExecutors.getInstance().networkIO(), AppExecutors.getInstance().mainThread())

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        Log.d(TAG, "onCreateView")
        initBinding(inflater, container)
        return mBinding.root
    }

    override fun onViewCreated(
            view: View,
            savedInstanceState: Bundle?
    ) {
        Log.d(TAG, "onViewCreated")
        super.onViewCreated(view, savedInstanceState)
        val args = AbilitiesFragmentArgs.fromBundle(requireArguments())
        mUserId = args.userId
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        Log.d(TAG, "onActivityCreated")
        super.onActivityCreated(savedInstanceState)
        initViewModel()
        initOperationsAdapter()
        setupBinding()
    }

    override fun onResume() {
        Log.d(TAG, "onResume")
        super.onResume()
        refreshAbilities()
        AccountSource.getAccount(
                requireActivity(),
                false
        ) {
            mViewModel.isOwn.set(it != null && it.name == mUserId.toString())
        }
    }

    private fun initOperationsAdapter() {
        Log.d(TAG, "initOperationsAdapter")
        mAbilitiesAdapter = AbilitiesAdapter(mViewModel.isOwn)
    }

    private fun initBinding(
            inflater: LayoutInflater,
            container: ViewGroup?
    ) {
        Log.d(TAG, "initBinding")
        mBinding = AbilitiesFragmentBinding.inflate(inflater, container, false)
    }

    private fun initViewModel() {
        Log.d(TAG, "initViewModel")
        mViewModel = ViewModelProvider(requireActivity()).get(AbilitiesViewModel::class.java)
    }

    private fun setupBinding() {
        Log.d(TAG, "setupBinding")
        mBinding.viewModel = mViewModel
        mBinding.list.recyclerView.adapter = mAbilitiesAdapter
        mBinding.refreshListener = SwipeRefreshLayout.OnRefreshListener {
            refreshAbilities()
        }
        mBinding.userActionListener = object : UserActionListener {
            override fun onAddAbilityClick() {
                val action: NavDirections = AbilitiesFragmentDirections.actionAbilitiesFragmentToAddAbilityFragment()
                NavHostFragment.findNavController(this@AbilitiesFragment).navigate(action)
            }
        }
    }

    private fun refreshAbilities() {
        Log.d(TAG, "refreshAbilities")
        mViewModel.downloadInProgress.set(true)
        mViewModel.abilities = mAsyncRepository.getLiveDataPagedListFromDataSource(AbilitiesDataSource.AbilitiesDataSourceFactory(mUserId))
        mViewModel.abilities?.observe(requireActivity()) { pagedList: PagedList<IAbility?>? ->
            mViewModel.isEmpty.set(pagedList?.isEmpty() ?: true)
            mAbilitiesAdapter.submitList(pagedList)
            mViewModel.downloadInProgress.set(false)
        }
    }
}